### ISSUE TITLE
Synaptic Intelligence implementation differs from the paper?

### ISSUE THREAD

**Issue by AndreaCossu** [ContinualAI/avalanche#106]
I was looking at Synaptic Intelligence implementation in [here](https://github.com/vlomonaco/avalanche/blob/master/avalanche/training/strategies/syn_int/utils.py). 

**Problem 1:**** line 103, when computing the cumulative trajectory (that is, the weight importance Big Omega), the parameter `c` is multiplied to the current trajectory (small omega in the paper). However, I think that, by following Eq. 4 in the paper, `c` should only be used as the hyperparameter weighting the loss and the penalty, like the `lambda` of EWC. In Eq. 5 in fact there is no `c` when computing cumulative trajectory.  In line 129 the penalty is weighted by `lambda`, so I am confused about it.  
**Proposed solutionI** I would remove the `c` in line 103 and substituting `lambda` with `c` in line 129.

**Problem 2:** The trajectory (small omega) is computed in line 95. The gradient taken in input however, is the gradient of the entire loss function (classification loss + Synaptic Intelligence penalty). In the original paper (Eq. 3 and explanation below) the gradient is taken with respect to the classification loss only.  
**Proposed solution** The only way to achieve this in our framework is to do all the computation in the `compute_loss`, by first computing classification loss, getting gradients (`retain_graph=True`), updating trajectory, add penalty, add gradient with respect to the penalty to the previous gradients and optimize parameters.

Do you agree with this or am I missing something?

**Comment by AndreaCossu** [comment]
I think that we can close this issue if we link the correct paper in the doc (I can take care of this). What do you say?

**Comment by vlomonaco** [comment]
Yes @AndreaCossu thanks!
